File: frontend/src/types/markets.ts
/**
 * Market calendar contracts. Mirrors `backend/app/schemas/markets.py`.
 *
 * The backend decides whether a market is open. This file describes the answer
 * it returns — no component computes a trading schedule.
 */

import type { AssetClass } from './instruments';

export type TradingStatus =
  | 'OPEN'
  | 'PRE_OPEN'
  | 'CLOSED'
  | 'WEEKEND'
  | 'HOLIDAY';

export interface MarketStatus {
  symbol: string;
  asset_class: AssetClass;
  /** Human label, e.g. `NSE Cash Market`. */
  market: string;
  /** Calendar governing it: `nse` or `crypto`. */
  calendar: string;
  status: TradingStatus;
  is_open: boolean;
  /** True for a market with no session boundaries at all. */
  is_24x7: boolean;
  timezone: string;
  server_time: string;
  /** Null for a 24/7 market — it has no next open. */
  next_open: string | null;
  /** Null for a 24/7 market — it has no next close. */
  next_close: string | null;
  reason: string;
  /**
   * Whether the backend REFUSES orders while closed. When false the status is
   * informational and trading still works, which is the default for this
   * paper-trading lab.
   */
  enforced: boolean;
}

export const STATUS_LABELS: Record<TradingStatus, string> = {
  OPEN: 'Open',
  PRE_OPEN: 'Pre-open',
  CLOSED: 'Closed',
  WEEKEND: 'Weekend',
  HOLIDAY: 'Holiday',
};

/** Badge tone per status. OPEN is the only positive one. */
export const STATUS_TONE: Record<TradingStatus, 'positive' | 'warning' | 'muted'> = {
  OPEN: 'positive',
  PRE_OPEN: 'warning',
  CLOSED: 'muted',
  WEEKEND: 'muted',
  HOLIDAY: 'muted',
};

File: backend/app/schemas/markets.py
"""Market calendar API contracts.

The backend resolves trading status; the frontend renders it. No React
component computes whether a market is open.
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.market_data.instruments import AssetClass
from app.markets.calendar import TradingStatus


class MarketStatusResponse(BaseModel):
    """Whether one instrument's market is open, and when it next changes."""

    model_config = ConfigDict(from_attributes=True)

    symbol: str
    asset_class: AssetClass
    market: str = Field(description='Human label, e.g. "NSE Cash Market".')
    calendar: str = Field(description='Calendar governing it: "nse" or "crypto".')
    status: TradingStatus = Field(
        description="OPEN, PRE_OPEN, CLOSED, WEEKEND or HOLIDAY."
    )
    is_open: bool = Field(description="Whether continuous trading is available now.")
    is_24x7: bool = Field(description="True for a market with no session boundaries.")
    timezone: str = Field(description="IANA zone the schedule is expressed in.")
    server_time: datetime = Field(description="The moment this describes.")
    next_open: datetime | None = Field(
        default=None,
        description="When trading next becomes available. Null for a 24/7 market.",
    )
    next_close: datetime | None = Field(
        default=None,
        description="When trading next stops. Null for a 24/7 market.",
    )
    reason: str = Field(description="Human explanation of the status.")
    enforced: bool = Field(
        description=(
            "Whether the trading engine REFUSES orders while closed "
            "(ENFORCE_MARKET_HOURS). When false the status is informational."
        )
    )
